cli/azd/pkg/async/task_test.go

```diff
@@ -0,0 +1,295 @@
+package async
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/test/mocks"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTaskWithResult(t *testing.T) {
+	expectedResult := "result"
+
+	task := NewTask[string]()
+	task.Run(func(ctx *TaskContext[string]) {
+		time.Sleep(250 * time.Millisecond)
+		ctx.SetResult(expectedResult)
+	})
+
+	actualResult, err := task.Await()
+
+	require.Equal(t, expectedResult, actualResult)
+	require.Nil(t, err)
+}
+
+func TestTaskWithAwait(t *testing.T) {
+	expectedResult := "result"
+
+	task := NewTask[string]()
+	task.Run(func(ctx *TaskContext[string]) {
+		time.Sleep(250 * time.Millisecond)
+		ctx.SetResult(expectedResult)
+	})
+
+	actualResult, err := task.Await()
+
+	require.Equal(t, expectedResult, actualResult)
+	require.Nil(t, err)
+}
+
+func TestTaskWithError(t *testing.T) {
+	expectedError := errors.New("example error")
+
+	task := NewTask[string]()
+	task.Run(func(ctx *TaskContext[string]) {
+		time.Sleep(250 * time.Millisecond)
+		ctx.SetError(expectedError)
+	})
+
+	actualResult, err := task.Await()
+
+	require.Equal(t, "", actualResult)
+	require.Equal(t, expectedError, err)
+}
+
+func TestTaskWithInvalidUsage(t *testing.T) {
+	require.Panics(t, func() {
+		task := NewTask[string]()
+		task.Run(func(ctx *TaskContext[string]) {
+			time.Sleep(250 * time.Millisecond)
+			ctx.SetError(errors.New("error"))
+			ctx.SetResult("value")
+		})
+
+		_, _ = task.Await()
+	})
+}
+
+func TestTaskWithProgressWithResult(t *testing.T) {
+	expectedResult := "result"
+	progress := []string{}
+
+	task := NewTaskWithProgress[string, string]()
+
+	go func() {
+		for status := range task.Progress() {
+			progress = append(progress, status)
+		}
+	}()
+
+	task.Run(func(ctx *TaskContextWithProgress[string, string]) {
+		ctx.SetProgress("thing 1")
+		time.Sleep(250 * time.Millisecond)
+		ctx.SetProgress("thing 2")
+		ctx.SetResult(expectedResult)
```

Comment:
What happen if the `func` never calls `SetResult`  or `SetError` ?

Reply:
Good candidate for the cancellation you spoke of.  I'm not going to cover every use case right now though.